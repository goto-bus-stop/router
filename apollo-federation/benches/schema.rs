use apollo_federation::schema::ValidFederationSchema;
use criterion::*;

fn bench_schema_setup(c: &mut Criterion) {
    let schema = apollo_compiler::Schema::parse_and_validate(
        include_str!("../../fuzz/supergraph-fed2.graphql"),
        "supergraph-fed2.graphql",
    )
    .unwrap();

    c.bench_function("schema_setup", move |b| {
        b.iter(|| {
            for _ in 0..1_000 {
                black_box(ValidFederationSchema::new(schema.clone()).unwrap());
            }
        });
    });
}

criterion_group!(schema_benches, bench_schema_setup);
criterion_main!(schema_benches);
