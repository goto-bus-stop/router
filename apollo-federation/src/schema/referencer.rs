use apollo_compiler::schema::Name;
use indexmap::IndexMap;
use indexmap::IndexSet;

use crate::error::FederationError;
use crate::error::SingleFederationError;
use crate::schema::position::DirectiveArgumentPosition;
use crate::schema::position::EnumPosition;
use crate::schema::position::EnumValuePosition;
use crate::schema::position::InputObjectFieldPosition;
use crate::schema::position::InputObjectPosition;
use crate::schema::position::InterfaceFieldArgumentPosition;
use crate::schema::position::InterfaceFieldPosition;
use crate::schema::position::InterfacePosition;
use crate::schema::position::ObjectFieldArgumentPosition;
use crate::schema::position::ObjectFieldPosition;
use crate::schema::position::ObjectPosition;
use crate::schema::position::ScalarPosition;
use crate::schema::position::SchemaDefinitionPosition;
use crate::schema::position::SchemaRootPosition;
use crate::schema::position::UnionPosition;
use crate::schema::position::UnionTypenamePosition;

#[derive(Debug, Clone, Default)]
pub(crate) struct Referencers {
    pub(crate) scalar_types: IndexMap<Name, ScalarTypeReferencers>,
    pub(crate) object_types: IndexMap<Name, ObjectTypeReferencers>,
    pub(crate) interface_types: IndexMap<Name, InterfaceTypeReferencers>,
    pub(crate) union_types: IndexMap<Name, UnionTypeReferencers>,
    pub(crate) enum_types: IndexMap<Name, EnumTypeReferencers>,
    pub(crate) input_object_types: IndexMap<Name, InputObjectTypeReferencers>,
    pub(crate) directives: IndexMap<Name, DirectiveReferencers>,
}

impl Referencers {
    pub(crate) fn contains_type_name(&self, name: &str) -> bool {
        self.scalar_types.contains_key(name)
            || self.object_types.contains_key(name)
            || self.interface_types.contains_key(name)
            || self.union_types.contains_key(name)
            || self.enum_types.contains_key(name)
            || self.input_object_types.contains_key(name)
    }

    pub(crate) fn get_scalar_type(
        &self,
        name: &str,
    ) -> Result<&ScalarTypeReferencers, FederationError> {
        self.scalar_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Scalar type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_object_type(
        &self,
        name: &str,
    ) -> Result<&ObjectTypeReferencers, FederationError> {
        self.object_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Object type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_interface_type(
        &self,
        name: &str,
    ) -> Result<&InterfaceTypeReferencers, FederationError> {
        self.interface_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Interface type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_union_type(
        &self,
        name: &str,
    ) -> Result<&UnionTypeReferencers, FederationError> {
        self.union_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Union type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_enum_type(
        &self,
        name: &str,
    ) -> Result<&EnumTypeReferencers, FederationError> {
        self.enum_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Enum type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_input_object_type(
        &self,
        name: &str,
    ) -> Result<&InputObjectTypeReferencers, FederationError> {
        self.input_object_types.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Input object type referencers unexpectedly missing type".to_owned(),
            }
            .into()
        })
    }

    pub(crate) fn get_directive(
        &self,
        name: &str,
    ) -> Result<&DirectiveReferencers, FederationError> {
        self.directives.get(name).ok_or_else(|| {
            SingleFederationError::Internal {
                message: "Directive referencers unexpectedly missing directive".to_owned(),
            }
            .into()
        })
    }
}

#[derive(Debug, Clone, Default)]
pub(crate) struct ScalarTypeReferencers {
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) object_field_arguments: IndexSet<ObjectFieldArgumentPosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
    pub(crate) interface_field_arguments: IndexSet<InterfaceFieldArgumentPosition>,
    pub(crate) union_fields: IndexSet<UnionTypenamePosition>,
    pub(crate) input_object_fields: IndexSet<InputObjectFieldPosition>,
    pub(crate) directive_arguments: IndexSet<DirectiveArgumentPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct ObjectTypeReferencers {
    pub(crate) schema_roots: IndexSet<SchemaRootPosition>,
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
    pub(crate) union_types: IndexSet<UnionPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct InterfaceTypeReferencers {
    pub(crate) object_types: IndexSet<ObjectPosition>,
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) interface_types: IndexSet<InterfacePosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct UnionTypeReferencers {
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct EnumTypeReferencers {
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) object_field_arguments: IndexSet<ObjectFieldArgumentPosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
    pub(crate) interface_field_arguments: IndexSet<InterfaceFieldArgumentPosition>,
    pub(crate) input_object_fields: IndexSet<InputObjectFieldPosition>,
    pub(crate) directive_arguments: IndexSet<DirectiveArgumentPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct InputObjectTypeReferencers {
    pub(crate) object_field_arguments: IndexSet<ObjectFieldArgumentPosition>,
    pub(crate) interface_field_arguments: IndexSet<InterfaceFieldArgumentPosition>,
    pub(crate) input_object_fields: IndexSet<InputObjectFieldPosition>,
    pub(crate) directive_arguments: IndexSet<DirectiveArgumentPosition>,
}

#[derive(Debug, Clone, Default)]
pub(crate) struct DirectiveReferencers {
    pub(crate) schema: Option<SchemaDefinitionPosition>,
    pub(crate) scalar_types: IndexSet<ScalarPosition>,
    pub(crate) object_types: IndexSet<ObjectPosition>,
    pub(crate) object_fields: IndexSet<ObjectFieldPosition>,
    pub(crate) object_field_arguments: IndexSet<ObjectFieldArgumentPosition>,
    pub(crate) interface_types: IndexSet<InterfacePosition>,
    pub(crate) interface_fields: IndexSet<InterfaceFieldPosition>,
    pub(crate) interface_field_arguments: IndexSet<InterfaceFieldArgumentPosition>,
    pub(crate) union_types: IndexSet<UnionPosition>,
    pub(crate) enum_types: IndexSet<EnumPosition>,
    pub(crate) enum_values: IndexSet<EnumValuePosition>,
    pub(crate) input_object_types: IndexSet<InputObjectPosition>,
    pub(crate) input_object_fields: IndexSet<InputObjectFieldPosition>,
    pub(crate) directive_arguments: IndexSet<DirectiveArgumentPosition>,
}
