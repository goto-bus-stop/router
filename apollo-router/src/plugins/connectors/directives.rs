#![allow(dead_code)]

use std::collections::HashMap;

use apollo_compiler::schema::{Component, Directive, EnumValueDefinition, Value};
use apollo_compiler::Node;

use super::selection_parser::Selection as JSONSelection;
use super::url_path_parser::Template as URLPathTemplate;

pub(super) const SOURCE_API_DIRECTIVE_NAME: &str = "source_api";
const HTTP_ARGUMENT_NAME: &str = "http";

#[derive(Debug)]
pub(super) struct SourceAPI {
    name: String,
    http: Option<HTTPSourceAPI>,
}

impl SourceAPI {
    pub(super) fn from_directive(name: String, component: &Component<EnumValueDefinition>) -> Self {
        let http = component
            .directives
            .0
            .iter()
            .find(|d| d.name == SOURCE_API_DIRECTIVE_NAME)
            .map(|directive| HTTPSourceAPI::from_directive(directive));
        Self { name, http }
    }
}

#[derive(Debug)]
pub(super) struct HTTPSourceAPI {
    base_url: String,
    default: Option<bool>,
    headers: Vec<HTTPHeaderMapping>,
}

impl HTTPSourceAPI {
    // todo: probably a result instead of unwraps ^^
    pub(super) fn from_directive(directive: &Node<Directive>) -> Self {
        let mut base_url = Default::default();
        let mut default = Default::default();
        let mut headers = Default::default();

        directive
            .arguments
            .iter()
            .find(|argument| argument.name == HTTP_ARGUMENT_NAME)
            // TODO: error handling plz ^^'
            .unwrap()
            .value
            .as_object()
            // TODO: error handling plz ^^'
            .unwrap()
            .iter()
            .for_each(|(name, node)| match name.as_str() {
                // TODO: error handling plz ^^'
                "base_url" => base_url = node.as_str().unwrap().to_string(),
                "default" => default = node.to_bool(),
                "headers" => headers = HTTPHeaderMapping::from_header_arguments(node),
                other => todo!("graceful error handling {other}"),
            });

        Self {
            base_url: base_url,
            default,
            headers,
        }
    }
}

#[derive(Debug)]
pub(super) struct HTTPHeaderMapping {
    name: String,
    r#as: Option<bool>,
    value: Option<String>,
}

impl HTTPHeaderMapping {
    // TODO: maybe a result?
    pub(super) fn from_header_arguments(argument: &Node<Value>) -> Vec<Self> {
        vec![]
    }
}

#[derive(Debug)]
pub(super) struct SourceType {
    api: String,
    http: Option<HTTPSourceType>,
    selection: Option<JSONSelection>,
    key_type_map: Option<KeyTypeMap>,
}

// TODO: impl tryfrom with XOR validation on methods
#[derive(Debug)]
pub(super) struct HTTPSourceType {
    get: Option<URLPathTemplate>,
    post: Option<URLPathTemplate>,
    headers: Vec<HTTPHeaderMapping>,
    body: Option<JSONSelection>,
}

#[derive(Debug)]
pub(super) struct KeyTypeMap {
    key: String,
    // Dictionary mapping possible __typename strings to values of the JSON
    // property named by key.
    type_map: HashMap<String, String>, // TODO: is this accurate?
}

#[derive(Debug)]
pub(super) struct SourceField {
    api: Option<String>,
    http: Option<HTTPSourceField>,
    selection: Option<JSONSelection>,
}

// TODO: impl tryfrom with XOR validation on methods
#[derive(Debug)]
pub(super) struct HTTPSourceField {
    get: Option<URLPathTemplate>,
    post: Option<URLPathTemplate>,
    put: Option<URLPathTemplate>,
    patch: Option<URLPathTemplate>,
    delete: Option<URLPathTemplate>,
    headers: Vec<HTTPHeaderMapping>,
    body: Option<JSONSelection>,
}
