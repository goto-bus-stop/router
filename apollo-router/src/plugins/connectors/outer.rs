use apollo_compiler::ast;
use apollo_compiler::ast::Definition;
use apollo_compiler::ast::Selection;
use apollo_compiler::Node;
use displaydoc::Display;
use futures::StreamExt;
use serde::Serialize;
use serde_json_bytes::ByteString;
use serde_json_bytes::Map;
use serde_json_bytes::Value;
use thiserror::Error;

use super::request_response::HackEntityResponseKey;
use crate::graphql;
use crate::services::SubgraphRequest;
use crate::services::SubgraphResponse;
use crate::services::SupergraphRequest;
use crate::Context;

const ENTITIES: &str = "_entities";
const REPRESENTATIONS: &str = "representations";
const TYPENAME: &str = "__typename";

pub(super) fn map_request(
    mut outer: SubgraphRequest,
) -> Result<SupergraphRequest, OuterRequestError> {
    if let Some((body, context)) =
        convert_entities_field_to_magic_finder(outer.subgraph_request.body_mut(), &outer.context)?
    {
        let mut supergraph_request = http::Request::builder()
            .method(outer.subgraph_request.method())
            .uri(outer.subgraph_request.uri())
            .version(outer.subgraph_request.version())
            .body(body)
            .map_err(|_todo| OuterRequestError::NewRequestBodyInvalid)?;
        *supergraph_request.headers_mut() = outer.subgraph_request.headers().clone();

        Ok(SupergraphRequest {
            supergraph_request,
            context,
        })
    } else {
        Ok(SupergraphRequest {
            supergraph_request: outer.subgraph_request,
            context: outer.context,
        })
    }
}

fn convert_entities_field_to_magic_finder(
    req: &mut graphql::Request,
    context: &Context,
) -> Result<Option<(graphql::Request, Context)>, OuterRequestError> {
    let query = req.query.clone().ok_or(OuterRequestError::QueryMissing)?;

    if let Some(entities_field) = find_entities_field(&query)? {
        let type_conditions = collect_type_conditions(&entities_field.selection_set);
        let typenames = collect_typenames_from_representations(&req.variables)?;

        // TODO check that all typenames are in type_conditions
        // TODO handle abstract types
        debug_assert!(type_conditions
            .iter()
            .all(|t| typenames.contains(&t.as_str())));

        let magic_finder_field = ast::Name::new(format!(
            "_{}_finder",
            typenames
                .first()
                .map(|s| (*s).to_string())
                .unwrap_or_default()
        ))
        .map_err(|_todo| OuterRequestError::InvalidMagicFinderFieldName)?;

        req.query = Some(query.clone().replace(ENTITIES, &magic_finder_field));

        context
            .private_entries
            .lock()
            .insert(HackEntityResponseKey(magic_finder_field.to_string()));

        Ok(Some((req.to_owned(), context.clone())))
    } else {
        Ok(None)
    }
}

fn find_entities_field(query: &str) -> Result<Option<Node<ast::Field>>, OuterRequestError> {
    let doc = apollo_compiler::ast::Document::parse(query.to_owned(), "op.graphql")
        .map_err(|_todo| OuterRequestError::QueryParseError)?;

    // Because this operation comes from the query planner, we can assume a single operation
    let op = doc
        .definitions
        .iter()
        .find_map(|def| match def {
            Definition::OperationDefinition(op) => Some(op),
            _ => None,
        })
        .ok_or(OuterRequestError::QueryMissing)?;

    Ok(op
        .selection_set
        .first()
        .and_then(|s| match s {
            Selection::Field(f) => {
                if f.name == ENTITIES {
                    Some(f)
                } else {
                    None
                }
            }
            _ => None,
        })
        .cloned())
}

fn collect_type_conditions(selection_set: &[Selection]) -> Vec<&ast::Name> {
    selection_set
        .iter()
        .filter_map(|s| match s {
            Selection::InlineFragment(f) => f.type_condition.as_ref(),
            _ => None,
        })
        .collect()
}

fn collect_typenames_from_representations(
    variables: &Map<ByteString, Value>,
) -> Result<Vec<&str>, OuterRequestError> {
    variables
        .get(REPRESENTATIONS)
        .ok_or(OuterRequestError::RepresentationsMissing)?
        .as_array()
        .ok_or(OuterRequestError::RepresentationsNotArray)?
        .iter()
        .map(|v| {
            v.as_object()
                .ok_or(OuterRequestError::ReperesentationNotObject)?
                .get(TYPENAME)
                .ok_or(OuterRequestError::TypenameMissing)?
                .as_str()
                .ok_or(OuterRequestError::TypenameNotString)
        })
        .collect::<Result<Vec<_>, _>>()
}

pub(super) async fn map_response(
    inner: crate::services::supergraph::Response,
    context: &crate::Context,
) -> Result<SubgraphResponse, OuterResponseError> {
    let hack_response_key = context
        .private_entries
        .lock()
        .remove::<HackEntityResponseKey>();

    if let Some(hack_response_key) = hack_response_key {
        let (parts, mut body) = inner.response.into_parts();

        // TODO multipart responses from connector subgraphs?
        let mut body = body
            .next()
            .await
            .ok_or(OuterResponseError::BadResponseBody)?;

        if let Some(ref mut data) = body.data {
            let data = data
                .as_object_mut()
                .ok_or(OuterResponseError::DataNotObject)?;

            if let Some(value) = data.remove(hack_response_key.0.as_str()) {
                data.insert(ENTITIES, value);
            }
        }

        Ok(SubgraphResponse {
            response: http::Response::from_parts(parts, body),
            context: context.clone(),
        })
    } else {
        // TODO do I have to do this if I don't need to touch the response?
        let (parts, mut body) = inner.response.into_parts();
        let body = body
            .next()
            .await
            .ok_or(OuterResponseError::BadResponseBody)?;
        Ok(SubgraphResponse {
            response: http::Response::from_parts(parts, body),
            context: context.clone(),
        })
    }
}

#[derive(Error, Display, Debug, Clone, Serialize, Eq, PartialEq)]
pub(super) enum OuterRequestError {
    /// Query is missing
    QueryMissing,
    /// Cannot parse query
    QueryParseError,
    /// Representations variable missing
    RepresentationsMissing,
    /// Respresentations is not an array
    RepresentationsNotArray,
    /// Representation is not an object
    ReperesentationNotObject,
    /// Typename missing
    TypenameMissing,
    /// Typename is not a string
    TypenameNotString,
    /// Invalid magic finder field name
    InvalidMagicFinderFieldName,
    /// Could not construct new GraphQL request body
    NewRequestBodyInvalid,
}

#[derive(Error, Display, Debug, Clone, Serialize, Eq, PartialEq)]
pub(super) enum OuterResponseError {
    /// Could not read response body
    BadResponseBody,
    /// Data is not an object
    DataNotObject,
}
