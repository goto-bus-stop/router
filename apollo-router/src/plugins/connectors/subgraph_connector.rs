//! Authorization plugin

// To remove once the commented code is out
#![allow(dead_code)]
#![allow(unused_imports)]
#![allow(clippy::needless_borrow)]
#![allow(clippy::extra_unused_lifetimes)]

use std::collections::HashMap;
use std::error::Error;
use std::ops::ControlFlow;
use std::ops::Deref;
use std::sync::Arc;

use apollo_compiler::schema::Directive;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::Value;
use apollo_compiler::Node;
use futures::Future;
use futures::StreamExt;
use http::uri::*;
use http::Uri;
use hyper::client::HttpConnector;
use hyper_rustls::ConfigBuilderExt;
use hyper_rustls::HttpsConnector;
use regex::Regex;
use tower::Layer;
use tower::ServiceBuilder;
use tower::ServiceExt;

use super::directives::HTTPSourceAPI;
use super::directives::SourceAPI;
use super::outer::map_request;
use super::outer::map_response;
use super::request_response::HackEntityResponseKey;
use super::Connector;
use crate::error::ConnectorDirectiveError;
use crate::error::FetchError;
use crate::layers::ServiceBuilderExt;
use crate::services::layers::query_analysis::QueryAnalysisLayer;
use crate::services::new_service::ServiceFactory;
use crate::services::subgraph;
use crate::services::trust_dns_connector::new_async_http_connector;
use crate::services::trust_dns_connector::AsyncHyperResolver;
use crate::services::MakeSubgraphService;
use crate::services::SupergraphCreator;
use crate::services::SupergraphRequest;
use crate::spec::Query;
use crate::spec::Schema;
use crate::spec::Selection;
use crate::Configuration;

#[derive(Clone)]
pub(crate) struct SubgraphConnector {
    creator: SupergraphCreator,
    query_analysis_layer: QueryAnalysisLayer,
}

impl SubgraphConnector {
    pub(crate) fn for_schema(
        schema: Arc<Schema>,
        configuration: Arc<Configuration>,
        creator: SupergraphCreator,
    ) -> Result<Self, ConnectorDirectiveError> {
        let query_analysis_layer = QueryAnalysisLayer::new(schema.clone(), configuration);
        Ok(Self {
            creator,
            query_analysis_layer,
        })
    }
}

use std::task::Poll;

use futures::future::BoxFuture;
use tower::BoxError;
use tower::Service;

use crate::services::SubgraphRequest;
use crate::services::SubgraphResponse;

/// "OUTER" SUBGRAPH -> "INNER" SUPERGRAPH SERVICE

impl tower::Service<SubgraphRequest> for SubgraphConnector {
    type Response = SubgraphResponse;
    type Error = BoxError;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, request: SubgraphRequest) -> Self::Future {
        let context = request.context.clone();
        let inner = match map_request(request) {
            Ok(inner) => inner,
            Err(e) => return Box::pin(async move { Err(BoxError::from(e)) }),
        };

        let query_analysis_layer = self.query_analysis_layer.clone();
        let service = self.creator.make_connector();

        Box::pin(async move {
            let res = match query_analysis_layer.supergraph_request(inner).await {
                Ok(req) => service.oneshot(req).await?,
                Err(res) => res,
            };

            map_response(res, &context).await.map_err(BoxError::from)
        })
    }
}

/// INNER (CONNECTOR) SUBGRAPH SERVICE

#[derive(Clone)]
pub(crate) struct HTTPConnector {
    schema: Arc<Schema>,
    connector: Connector,
    client: hyper::Client<HttpsConnector<HttpConnector<AsyncHyperResolver>>>,
}

impl HTTPConnector {
    pub(crate) fn new(schema: Arc<Schema>, connector: Connector) -> Result<Self, BoxError> {
        let mut http_connector = new_async_http_connector()?;
        http_connector.set_nodelay(true);
        http_connector.set_keepalive(Some(std::time::Duration::from_secs(60)));
        http_connector.enforce_http(false);

        let tls_config = rustls::ClientConfig::builder()
            .with_safe_defaults()
            .with_native_roots()
            .with_no_client_auth();

        let http_connector = hyper_rustls::HttpsConnectorBuilder::new()
            .with_tls_config(tls_config)
            .https_or_http()
            .enable_http1()
            .enable_http2()
            .wrap_connector(http_connector);
        //TODO: add decompression
        let client = hyper::Client::builder().build(http_connector);

        Ok(Self {
            schema,
            connector,
            client,
        })
    }
}

impl<S> Layer<S> for HTTPConnector
where
    S: Clone,
{
    type Service = HTTPConnectorService<S>;

    fn layer(&self, inner: S) -> Self::Service {
        HTTPConnectorService {
            inner,
            schema: self.schema.clone(),
            connector: self.connector.clone(),
            client: self.client.clone(),
        }
    }
}

#[derive(Clone)]
pub(crate) struct HTTPConnectorService<S> {
    inner: S,
    schema: Arc<Schema>,
    connector: Connector,
    client: hyper::Client<HttpsConnector<HttpConnector<AsyncHyperResolver>>>,
}

impl<S> tower::Service<SubgraphRequest> for HTTPConnectorService<S>
where
    S: Service<SubgraphRequest>,
    S::Future: Future<Output = Result<SubgraphResponse, BoxError>> + Send + 'static,
{
    type Response = SubgraphResponse;
    type Error = BoxError;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner
            .poll_ready(cx)
            .map_err(|_| BoxError::from("http service is not ready"))
    }

    fn call(&mut self, request: SubgraphRequest) -> Self::Future {
        // dbg!(&request.subgraph_request, &request.subgraph_name);
        let schema = self.schema.clone();
        let connector = self.connector.clone();
        let context = request.context.clone();
        // dbg!(&connector);

        let client = self.client.clone();

        Box::pin(async move {
            let requests =
                connector.create_requests(request, Arc::from(schema.definitions.clone()))?;

            let tasks = requests.into_iter().map(|(req, res_params)| async {
                println!("HTTP {} {}", &req.method(), &req.uri());
                let mut res = client.request(req).await?;
                res.extensions_mut().insert(res_params);
                Ok::<_, BoxError>(res)
            });

            let results = futures::future::try_join_all(tasks).await;

            let responses = match results {
                Ok(responses) => responses,
                Err(e) => return Err(BoxError::from(e)),
            };

            let subgraph_response = connector.map_http_responses(responses, context).await?;
            dbg!(&subgraph_response.response.body().data);

            Ok(subgraph_response)

            // TODO: consider removing inner,
            // unless we have a nice subgraph HTTP service sometimes
            //
            // let fut = self.inner.call(request);
            // fut.await.map_err(BoxError::from)
        })
    }
}

#[cfg(test)]
mod tests {
    use std::convert::Infallible;
    use std::net::SocketAddr;
    use std::net::TcpListener;
    use std::sync::Arc;

    use http::header::CONTENT_TYPE;
    use http::Method;
    use http::StatusCode;
    use hyper::service::make_service_fn;
    use hyper::service::service_fn;
    use hyper::Body;
    use hyper::Server;
    use mime::APPLICATION_JSON;

    use super::*;
    use crate::router_factory::YamlRouterFactory;
    use crate::services::supergraph;
    use crate::TestHarness;

    const SCHEMA: &str = include_str!("../../../../examples/connectors/starstuff.graphql");

    async fn emulate_rest_connector(listener: TcpListener) {
        async fn handle(
            request: http::Request<Body>,
        ) -> Result<http::Response<String>, Infallible> {
            /*
                        type IP
              @join__type(graph: NETWORK, key: "ip")
              @sourceType(
                graph: "network"
                api: "ipinfo"
                http: { GET: "/json" }
                selection: "ip hostname city region country loc org postal timezone readme"
              ) {
              ip: ID!
              hostname: String
              city: String
              region: String
              country: String
              loc: String
              org: String
              postal: String
              timezone: String
              readme: String
            }
                         */

            let res = if request.method() == Method::GET
            /*&& request.uri().path() == "/json"*/
            {
                let value = serde_json::json! {{
                    "ip": "1.2.3.4",
                    "hostname": "hello",
                    "city": "Paris",
                    "region": "Ile de France",
                    "country": "France",
                    "loc": "1",
                    "org": "a",
                    "postal": "75000",
                    "timezone": "CEST",
                    "readme": "readme"
                }};
                Ok(http::Response::builder()
                    .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
                    .status(StatusCode::OK)
                    .body(serde_json::to_string(&value).expect("always valid"))
                    .unwrap())
            } else {
                Ok(http::Response::builder()
                    .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
                    .status(StatusCode::BAD_REQUEST)
                    .body(String::new())
                    .unwrap())
            };
            println!("generated service response: {res:?}");
            res
        }

        let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(handle)) });
        let server = Server::from_tcp(listener).unwrap().serve(make_svc);
        server.await.unwrap();
    }

    #[tokio::test]
    async fn nullability_formatting() {
        let listener = TcpListener::bind(SocketAddr::from(([127, 0, 0, 1], 0))).unwrap();
        let address = listener.local_addr().unwrap();
        let _spawned_task = tokio::task::spawn(emulate_rest_connector(listener));

        let schema = SCHEMA.replace(
            "https://ipinfo.io/",
            &format!("http://127.0.0.1:{}/", address.port()),
        );

        // we cannot use Testharness because the subgraph connectors are actually extracted in YamlRouterFactory
        let mut factory = YamlRouterFactory;
        use crate::router_factory::RouterSuperServiceFactory;
        let router_creator = factory
            .create(
                Arc::new(
                    serde_json::from_value(serde_json::json!({
                        "include_subgraph_errors": { "all": true }
                    }))
                    .unwrap(),
                ),
                schema,
                None,
                None,
            )
            .await
            .unwrap();
        let service = router_creator.create();

        let request = supergraph::Request::fake_builder()
            .query("query { serverNetworkInfo { ip city country } }")
            // Request building here
            .build()
            .unwrap()
            .try_into()
            .unwrap();
        let response = service
            .oneshot(request)
            .await
            .unwrap()
            .next_response()
            .await
            .unwrap()
            .unwrap();
        let response: serde_json::Value = serde_json::from_slice(&response).unwrap();

        insta::assert_json_snapshot!(response);
    }
}
