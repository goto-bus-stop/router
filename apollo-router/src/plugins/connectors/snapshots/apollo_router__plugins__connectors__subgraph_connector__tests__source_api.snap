---
source: apollo-router/src/plugins/connectors/subgraph_connector.rs
expression: source_apis_from_schema
---
{
    "CONTACTS": SourceAPI {
        name: "CONTACTS",
        http: Some(
            HTTPSourceAPI {
                base_url: "http://localhost:4002/contacts/",
                default: Some(
                    true,
                ),
                headers: [],
            },
        ),
    },
    "NOTES": SourceAPI {
        name: "NOTES",
        http: Some(
            HTTPSourceAPI {
                base_url: "http://localhost:4002/notes/",
                default: None,
                headers: [],
            },
        ),
    },
    "LEGACY_CONTACTS": SourceAPI {
        name: "LEGACY_CONTACTS",
        http: Some(
            HTTPSourceAPI {
                base_url: "http://localhost:4002/legacy/contacts/",
                default: None,
                headers: [],
            },
        ),
    },
}
